kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -42,9 +39,322 @@ import cats.syntax.either._
+    executorService.schedule(
+      new Thread {
+        override def run(): Unit = {
+          val lastFlushed = EventStorage.getLastFlushTime
+          val currentTime = System.currentTimeMillis()
+          if (currentTime - lastFlushed >= TimeThreshold) {
+            EventStorage.flush()
+            scheduleFlush(TimeThreshold)
+          } else {
+            scheduleFlush(TimeThreshold + lastFlushed - currentTime)
+          }
+        }
+      },
+      interval,
+      MILLISECONDS
+    )
+    ()
+  }
+
+  object EventStorage {
+    private val storedEvents              = ListBuffer.empty[Events]
+    private var byteCount                 = 0L
+    @volatile private var lastFlushedTime = 0L
+
+    def store(event: Array[Byte], key: String): Unit = {
+      val eventBytes = ByteBuffer.wrap(event)
+      val eventSize  = eventBytes.capacity
+      if (eventSize >= MaxBytes) {
+        log.error(
+          s"Record of size $eventSize bytes is too large - must be less than $MaxBytes bytes"
+        )
+      } else {
+        synchronized {
+          if (storedEvents.size + 1 > RecordThreshold || byteCount + eventSize > ByteThreshold) {
+            flush()
+          }
+          storedEvents += Events(eventBytes.array(), key)
+          byteCount += eventSize
+        }
+      }
+    }
+
+    def flush(): Unit = {
+      val eventsToSend = synchronized {
+        val evts = storedEvents.result
+        storedEvents.clear()
+        byteCount = 0
+        evts
+      }
+      lastFlushedTime = System.currentTimeMillis()
+      sinkBatch(eventsToSend.toList, KinesisStream(maybeSqs))(10)
+    }
+
+    def getLastFlushTime: Long = lastFlushedTime
+  }
+
+  def storeRawEvents(events: List[Array[Byte]], key: String): List[Array[Byte]] = {
+    events.foreach(e => EventStorage.store(e, key))
+    Nil
+  }
+
+  def scheduleBatch(batch: List[Events], lastBackoff: Long = minBackoff): Unit = {
+    val nextBackoff = getNextBackoff(lastBackoff)
+    executorService.schedule(new Thread {
+      override def run(): Unit =
+        sinkBatch(batch, KinesisStream(maybeSqs), nextBackoff)(10)
+    }, lastBackoff, MILLISECONDS)
+    ()
+  }
+
+  /**
+    * The number of retries is unlimited, but targets switch every 10 attempts if an SQS buffer is configured.
+    * @param batch A batch of events to be written
+    * @param target A Kinesis stream or SQS queue to write events to
+    * @param nextBackoff Period of time to wait before next attempt if retrying
+    * @param retriesLeft How many retries are left to the current target. Once exhausted, the target will be switched.
+    */
+  def sinkBatch(batch: List[Events], target: Target, nextBackoff: Long = minBackoff)(
+    retriesLeft: Int                                                   = 10
+  ): Unit =
+    if (batch.nonEmpty) target match {
+      case KinesisStream(buffer) =>
+        log.info(s"Writing ${batch.size} Thrift records to Kinesis stream $streamName.")
```

Comment:
I don't think that we need to add the name of the stream to each line of log

Reply:
It's very helpful in testing though. :)